[Security Solution][Endpoint] Accept all kinds of filenames (without wildcard) in wildcard-ed event filter `file.path.text`

ashokaditya opened the pull request:
## Summary

~Fixes the filename regex to extract filenames from a wildcard file path.~
~The regex is updated to a much simpler pattern that covers a whole gamut file name pattern.~
~Also, it's ReDoS safe. You can also test it [here](https://devina.io/redos-checker)~
- [x] Updated the logic to verify filenames without wildcards (`?` or`*`).
- [x] Updated `file.path.text` artifact entry to exclude `.text`.
- [x] Updated/added test for above.


~screenshots~(outdated)
![Screenshot 2022-03-22 at 09 35 55](https://user-images.githubusercontent.com/1849116/159442124-d2849b90-311f-46ff-9f8c-5f78c35fa4d2.png)~
![Screenshot 2022-03-22 at 09 29 17](https://user-images.githubusercontent.com/1849116/159442131-1a0a4c54-5117-4c35-bd0c-033633ba3bf4.png)



related:
elastic/kibana/pull/125202

**Note:**
We decided that we don't want to add the additional optimized `file.name` or `process.name` entry for wildcard paths if such an entry already exists (is added by the user). Here is an example entry of an event filter and generated artifact. Currently, it is not possible to add a wildcard entry for `process.executable.caseless` in event filters, but one can add such an entry in trusted apps. However, a test is included that covers this case.

For a filter such as this
![Screenshot 2022-03-23 at 10 44 01](https://user-images.githubusercontent.com/1849116/159670495-a29b29b8-4476-43b3-867c-ca60837083e2.png)

The generated artifact looks like this:
```json5
{
  "entries": [
    {
      "type": "simple",
      "entries": [
        {
          "field": "file.path",
          "operator": "included",
          "type": "exact_cased",
          "value": "/usr/lib/some.app"
        },
        {
          "field": "file.path",
          "operator": "included",
          "type": "wildcard_cased",
          "value": "/usr/*/some.app"
        },
        {
          "field": "file.path",
          "operator": "included",
          "type": "exact_cased_any",
          "value": [
            "/opt/bin/file.exe",
            "/usr/lib/some.app"
          ]
        },
        {
          "field": "file.name",
          "operator": "included",
          "type": "exact_cased",
          "value": "some.app"
        },
        {
          "field": "process.name",
          "operator": "included",
          "type": "exact_cased",
          "value": "/usr/lib/some-app.exe"
        },
        {
          "field": "process.name",
          "operator": "included",
          "type": "exact_cased",
          "value": "some-app.exe"
        }
      ]
    }
  ]
}
```

**Updated**
For event filters such as these:
![Screenshot 2022-03-23 at 10 24 44](https://user-images.githubusercontent.com/1849116/159666398-e665bec8-9a50-4d89-9ace-16f97c226c76.png)

the generated artifact looks like this:

```json5
{
  "entries": [
    {
      "type": "simple",
      "entries": [
        {
          "field": "file.path",
          "operator": "included",
          "type": "wildcard_caseless",
          "value": "C:\\*\\*\\filter-this-doc.txt"
        },
        {
          "field": "file.name",
          "operator": "included",
          "type": "exact_caseless",
          "value": "filter-this-doc.txt"
        }
      ]
    },
    {
      "type": "simple",
      "entries": [
        {
          "field": "file.path",
          "operator": "included",
          "type": "wildcard_caseless",
          "value": "C:\\Users\\User\\*\\event-filter-test-file.txt"
        },
        {
          "field": "file.name",
          "operator": "included",
          "type": "exact_caseless",
          "value": "event-filter-test-file.txt"
        }
      ]
    }
  ]
}
```



### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

ashokaditya committed:
update filename regex to include multiple hyphens and periods

Uses a much simpler pattern that covers a whole gamut file name patterns.
fixes elastic/security-team/issues/3294

ashokaditya committed:
remove duplicated code

ashokaditya committed:
add tests for `process.name` entry for filenames with wildcard path

refs
elastic/kibana/pull/120349
elastic/kibana/pull/125202

elasticmachine commented:
Pinging @elastic/security-onboarding-and-lifecycle-mgt (Team:Onboarding and Lifecycle Mgt)

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

kevinlog commented on x-pack/plugins/security_solution/server/endpoint/lib/artifacts/lists.test.ts:
```diff
@@ -513,6 +513,358 @@ describe('artifacts lists', () => {
+              value,
+            },
+            {
+              field: 'process.name',
```
@ashokaditya when we are using `file.path.text` we cannot use `process.name` as it will almost always create an invalid rule since the `process.name` in the doc will be different.

Instead, we can add `file.name` when using the `file.path.text` field.  This will create a valid rule similarly as when we use `process.executable` and `process.name` together.

ashokaditya replied:
done 5bfd726acd459eacba783e356d08844c73a06fd2

ashokaditya committed:
Add file.name optimized entry when wildcard filepath in file.path.text has a filename

fixes elastic/security-team/issues/3294

FrankHassanabad reviewed (approved):
LGTM, thanks for the redos tests with the regex. Briefly looked through but we have a lot of trust here with these great PR's! All looks neat and easily readable.

intxgo commented on packages/kbn-securitysolution-utils/src/path_validations/index.ts:
```diff
@@ -37,11 +37,15 @@ export type TrustedAppEntryTypes = 'match' | 'wildcard';
  */
-export const WIN_EXEC_PATH = /(\\[-\w]+|\\[-\w]+[\.]+[\w]+)$/i;
-export const UNIX_EXEC_PATH = /(\/[-\w]+|\/[-\w]+[\.]+[\w]+)$/i;
+export const WIN_EXEC_PATH = /(\\[-\.\w ]+)$/i;
```
I'm not expert in regex, but does `\w` match Unicode and special ASCII codes above 128?
I've found the explanation `\w matches any word character (equivalent to [a-zA-Z0-9_])`
Windows file name can contain a lot of more characters, but also it cannot contain certain things like using two consecutive dots `..`
https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file

I have no opinion of how fool-proof it should be, as this is only to extract the file name part from the given path. I'd even say that it should suffice to extract just the string after the last `\`, provided it's non-empty.

ashokaditya replied:
This regex is actually not to extract the filename but to check if string after the last `\` (in windows) or `/` (in a unix system) has no wildcards. Currently, we use node's built-in `path` to extract the filenames if the filename is without a wildcard.

I see your point about the Unicode chars. I'll see what I can whip up to cover those cases.

intxgo replied:
what about doing what you mean `/(\\[^*?]+)$/i`

ashokaditya replied:
Thanks for looking into this @intxgo . 

`/(\\[^*?]+)$/i` works but unfortunately is not [ReDos safe](https://devina.io/redos-checker). 

I came up with the following which is safer.

`/(\\[-\.\w \p{L}]+)$/iu`

**Question:**
Should we cover this kind of naming also for UNIX systems?

ashokaditya replied:
@intxgo done 7e1a634013a657d9a80e4bd613dea6a71e62bd77

intxgo replied:
Don't forget about other valid characters
```
c:\anything\*\file.exe-2.bck
c:\anything\*\file.exe-?2.bck
c:\anything\*\file.exe*2.bck
c:\anything\*\file.exe-2.bck您正在访问的域名可以转
c:\anything\*\test.txt
c:\anything\*\test!.txt
c:\anything\*\test!#.txt
c:\anything\*\test~1.txt
c:\anything\*\test$.txt
```
... etc

Sure, Linux also allows many special ASCII characters inside paths, as well as Unicode.

intxgo replied:
on a second though, why not extracting the match with `/\\(.+)$/i` and then just do a simple linear find for '*' and '?'. That's get you covered and is ReDos safe.

intxgo replied:
@ashokaditya Limiting the string length would mitigate ReDos. There could be many FS on Linux, but it seems most of them keep a sensible file name length limit ~260 characters. I'll leave this research/decision to you. I think it's far less probable to deal with an FS allowing very long file names than to have file name with non `\w` characters in it, thus if you take out last ~300 characters from the path to check with regex it should be fine.

ashokaditya replied:
8161223cd22622c50111f121ca46a12290ef0d83

ashokaditya committed:
update regex to include unicode chars

review changes

elasticmachine commented:
Pinging @elastic/apm-ui (Team:apm)

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

ashokaditya committed:
add tests for `file.name` and `process.name` entries if it already exists

This works out of the box and we don't add endpoint related `file.name` or `process.name` entry when it already is added by the user

refs
elastic/kibana/pull/127958#discussion_r829086447
elastic/security-team/issues/3199

ashokaditya committed:
Merge branch 'main' into fix/olm-wildcard-filepath-filename-3294

ashokaditya committed:
fix `file.name` and `file.path.text` entries for linux and mac/linux

refs elastic/kibana/pull/127098

ashokaditya committed:
do not add endpoint optimized entry

Add `file.name` and `process.name` entry for wildcard path values only when file.name and process.name entries do not already exist.

The earlier commit 8a516ae9c0580eb44b57666e7a5934c543c3e4bb was mistakenly labeled as this worked out of the box. In the same commit we notice that the test data had a wildcard file path that did not add a `file.name` or `process.name` entry.

For more see:
elastic/kibana/pull/127958#discussion_r829086447
elastic/security-team/issues/3199

ashokaditya committed:
Merge branch 'main' into fix/olm-wildcard-filepath-filename-3294

ashokaditya committed:
Merge branch 'main' into fix/olm-wildcard-filepath-filename-3294

ashokaditya committed:
update regex to include gamut of unicode characters

review suggestions

ashokaditya committed:
remove regex altogether

simplifies the logic to check if path is without wildcard characters. This way it includes all other strings as valid filenames that do not have * or ?

ashokaditya committed:
Merge branch 'main' into fix/olm-wildcard-filepath-filename-3294

ashokaditya committed:
update artifact creation for `file.path.text` entries

Similar to when we normalize `file.path.caseless` entries, except that the `type` is `*_cased` for linux and `*_caseless` for non-linux

ashokaditya committed:
Merge branch 'main' into fix/olm-wildcard-filepath-filename-3294

kevinlog commented:
Checked it out and tried it - LGTM!

I added a matches event filter for a text file, `kevin.txt` on my connected VM.  I was not seeing events stream during the time the filter was active.  When I removed the filter from the Policy, the event for the file started streaming again.

See the gap in events between 8:05 and 8:18
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/56395104/159697864-9e1d5ae1-08f3-4145-b1de-d3183b825ac0.png">

The filter I used:
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/56395104/159697909-789c5fe0-f88e-4d9e-aa7b-796156252bc3.png">

ashokaditya commented:
@kevinlog thanks for checking this out. I forgot to mention that I checked this also without assigning the filter to a policy and that also works as expected.

paul-tavares reviewed (approved):
LGTM 👍

ashokaditya committed:
Merge branch 'main' into fix/olm-wildcard-filepath-filename-3294

ashokaditya committed:
Merge branch 'main' into fix/olm-wildcard-filepath-filename-3294

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 127432` or prevent reminders by adding the `backport:skip` label.
